### PR TITLE
feat: Add dryrun feature to Dynamo paths

### LINKED ISSUE #2081: ✨[Feature] Compilation Dry-Run in Dynamo
# Context
While many key `aten` converters are under development, the end-to-end performance of certain key models is greatly affected by segmentation. For instance, with `min_block_size=5` we obtain nearly 100 distinct TRT Engines in the graph, as of the current `main`. This is an unreasonable amount of segmentation, which slows down inference. A temporary solution to this is to increase the block size via `min_block_size`, however this has the implication that every converter benefits the same from acceleration, which is not necessarily the case. For instance, `convolution` may benefit from conversion much more than `add` or `sub`.

# Feature Proposal
In the converter registry, each converter can be assigned a weight based on the operation it is converting and the relative speedup which it would gain from conversion. For instance, `convolution` could have a weight of 5, while `add` has a weight of 1. From here, there are a few options for picking better segments:

- Ensure the weighted sum of the converter weights is at least `min_block_size`
- Remove `min_block_size` in favor of a simpler heuristic like `allowed_segmentation - {"none", "minimal", "moderate", "large", "most"}`

### LINKED ISSUE #2413: ✨[Feature] Dry-Run Functionality for Dynamo
# Context
In the Dynamo paths, there is a great deal of information behind-the-scenes, which could be helpful to users outside of "debug" mode. Specifically, users may want to get a glimpse of the operator coverage for their model, the structure of the model graph, or potentially the composition of each partition in the TRT/Torch split. These usecases can all be enabled via the Dry-Run functionality.

# Feature Proposal
A new compilation argument, `dry_run`, which when enabled, proceeds through all the steps of compilation through partitioning, stopping before conversion. In the `ir="dynamo"` path, the program would halt here, whereas in the `ir="toch_compile"` path, we would return the Torch graph so as to not cause an error in the `torch.compile` interpreter.

The `dry_run` argument could print (at least) the following valuable information and statistics:

- List of supported and unsupported operators
- Number of graphs in the network after partitioning
- Average number of TRT operators per segmented graph
- Schematic of the graph compositions, roughly:
```
Inputs --> (45 operators, TRT) ---> (23 operators, Torch) ---> (30 operators, TRT) --> Outputs
```
- Automatic recommendations for `min_block_size` choices to reduce (or increase) segmentation, such as:
```
For the minimal amount of segmentation, consider {min_block_size=45}, which will result in {2} graph segments.
For a reasonable amount of segmentation, consider {min_block_size=30}, which will result in {3} graph segments.
For the maximal amount of TRT-run-operators and most segmentation, consider {min_block_size=5}, which will result in {10} graph segments.
```
- I/O schematic for the graph, roughly:
```
Inputs  --(2, 3), (3, 4)-> (TRT) --(3, 7), (6, 3)-> (23 operators, Torch) --(5, 3)-> (30 operators, TRT) -(24, 1)-> Outputs
```
- **(Future)** Use `torch._dynamo.explain` to get graph break information from within the compiler, providing the number of graph breaks, and reasons for those breaks, in the Dry-Run outputs as well

### PR DESCRIPTION
**Pull request by gs-olive**
# Description
- Enables building of TRT engines with "dryrun" capabilities, meaning all of the phases except conversion are run and verbose logs of the graph structure and composition are printed for the user
- Improves general-purpose debug logging by printing dryrun stats to the debug logs regardless of option specification
- Provides intuitive schematic of the graph engines, inputs, and code path through the course of the graph

<details>
<summary>
<b>Sample Schematic</b>
</summary>

```python
++++++++++++++++++++++++++++++++++++++++++++++++++ Dry-Run Results for Graph ++++++++++++++++++++++++++++++++++++++++++++++++++

The graph consists of 89 Total Operators, of which 81 operators are supported, 91.01% coverage

The following ops are currently unsupported or excluded from conversion, and are listed with their op-count in the graph:
 torch.ops.aten.add.Tensor: 8

The following nodes are currently set to run in Torch:
Node: torch.ops.aten.add.Tensor, with layer location: /layer1/0/add
Node: torch.ops.aten.add.Tensor, with layer location: /layer1/1/add_1
Node: torch.ops.aten.add.Tensor, with layer location: /layer2/0/add_2
Node: torch.ops.aten.add.Tensor, with layer location: /layer2/1/add_3
Node: torch.ops.aten.add.Tensor, with layer location: /layer3/0/add_4
Node: torch.ops.aten.add.Tensor, with layer location: /layer3/1/add_5
Node: torch.ops.aten.add.Tensor, with layer location: /layer4/0/add_6
Node: torch.ops.aten.add.Tensor, with layer location: /layer4/1/add_7
Note: Some of the above nodes may be supported, but were not included in a TRT graph by the partitioner

Compiled with: CompilationSettings(precision=torch.float32, debug=True, workspace_size=0, min_block_size=1, torch_executed_ops={'torch.ops.aten.add.Tensor'}, pass_through_build_failures=False, max_aux_streams=None, version_compatible=False, optimization_level=None, use_python_runtime=False, truncate_long_and_double=False, use_fast_partitioner=False, enable_experimental_decompositions=False, device=Device(type=DeviceType.GPU, gpu_id=0), require_full_compilation=False, dryrun=True)

  Graph Structure:

   Inputs: List[Tensor: (1, 3, 224, 224)@float32]
    ...
    TRT Engine #1 - Submodule name: fused_0
     Engine Inputs: List[Tensor: (1, 512, 7, 7)@float32]
     Number of Operators in Engine: 4
     Engine Outputs: Tensor: (1, 1000)@float32
    ...
    TRT Engine #2 - Submodule name: fused_1
     Engine Inputs: List[Tensor: (1, 512, 7, 7)@float32]
     Number of Operators in Engine: 8
     Engine Outputs: Tuple(Tensor: (1, 512, 7, 7)@float32, Tensor: (1, 512, 7, 7)@float32)
    ...
    TRT Engine #3 - Submodule name: fused_2
     Engine Inputs: List[Tensor: (1, 256, 14, 14)@float32]
     Number of Operators in Engine: 11
     Engine Outputs: Tuple(Tensor: (1, 512, 7, 7)@float32, Tensor: (1, 512, 7, 7)@float32)
    ...
    TRT Engine #4 - Submodule name: fused_3
     Engine Inputs: List[Tensor: (1, 256, 14, 14)@float32]
     Number of Operators in Engine: 8
     Engine Outputs: Tuple(Tensor: (1, 256, 14, 14)@float32, Tensor: (1, 256, 14, 14)@float32)
    ...
    TRT Engine #5 - Submodule name: fused_4
     Engine Inputs: List[Tensor: (1, 128, 28, 28)@float32]
     Number of Operators in Engine: 11
     Engine Outputs: Tuple(Tensor: (1, 256, 14, 14)@float32, Tensor: (1, 256, 14, 14)@float32)
    ...
    TRT Engine #6 - Submodule name: fused_5
     Engine Inputs: List[Tensor: (1, 128, 28, 28)@float32]
     Number of Operators in Engine: 8
     Engine Outputs: Tuple(Tensor: (1, 128, 28, 28)@float32, Tensor: (1, 128, 28, 28)@float32)
    ...
    TRT Engine #7 - Submodule name: fused_6
     Engine Inputs: List[Tensor: (1, 64, 56, 56)@float32]
     Number of Operators in Engine: 11
     Engine Outputs: Tuple(Tensor: (1, 128, 28, 28)@float32, Tensor: (1, 128, 28, 28)@float32)
    ...
    TRT Engine #8 - Submodule name: fused_7
     Engine Inputs: List[Tensor: (1, 64, 56, 56)@float32]
     Number of Operators in Engine: 8
     Engine Outputs: Tuple(Tensor: (1, 64, 56, 56)@float32, Tensor: (1, 64, 56, 56)@float32)
    ...
    TRT Engine #9 - Submodule name: fused_8
     Engine Inputs: List[Tensor: (1, 3, 224, 224)@float32]
     Number of Operators in Engine: 12
     Engine Outputs: Tuple(Tensor: (1, 64, 56, 56)@float32, Tensor: (1, 64, 56, 56)@float32)
    ...
   Outputs: List[Tensor: (1, 1000)@float32]

  ------------------------- Aggregate Stats -------------------------

   Average Number of Operators per TRT Engine: 9.0
   Most Operators in a TRT Engine: 12

  ********** Recommendations **********

   - For minimal graph segmentation, select min_block_size=12 which would generate 1 TRT engine(s)
   - For moderate graph segmentation, select min_block_size=9 which would generate 4 TRT engine(s)
   - The current level of graph segmentation is equivalent to selecting min_block_size=4 which generates 9 TRT engine(s)
```
</details>

Fixes #2081
Fixes #2413

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ - ] I have added tests to verify my fix or my feature
  - Validated via CI
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Comment by narendasan** [comment]
Can you add a summary of user settings? just dump the struct?

**Comment by narendasan** [comment]
Also a list of ops to be run in PyTorch

**Comment by narendasan** [comment]
Also this (if it doesnt add to compilation time) should be added as debugging logging to all compilation calls. And if its printed at INFO level it might get masked depending on the users logging settings, I think if people are explicitly calling dry run it should be printed out in STDOUT

**Review by narendasan — APPROVED** [review]
LGTM, notes for later:

We should pretty print this Compiled with: `CompilationSettings(precision=torch.float32, debug=True, workspace_size=0, min_block_size=1, torch_executed_ops={'torch.ops.aten.add.Tensor'}, pass_through_build_failures=False, max_aux_streams=None, version_compatible=False, optimization_level=None, use_python_runtime=False, truncate_long_and_double=False, use_fast_partitioner=False, enable_experimental_decompositions=False, device=Device(type=DeviceType.GPU, gpu_id=0), require_full_compilation=False, dryrun=True)`

Also are we able to see the intermediate torch graphs just like we see the tensorrt ones?

**Comment by gs-olive** [comment]
@narendasan - thanks for the comments. I have added issues for the suggested improvements here: #2548. We are not able to see inputs to intermediate Torch graphs yet, because they are not always packaged as modules. Specifically, our global partitioner which uses Torch partitioning utilities, only packages the TRT engines as modules and leaves the Torch operators alone, which makes it difficult to group those into subgraphs. This improvement can be added for the fast partitioner, however, and this is noted in the new feature request issue.

**Comment by narendasan** [comment]
Can you not iterate across the graph and just make lists of non tensorrt ops, even just an idea of what is left out I think is probably some of the more important information this feature could produce?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 1]
```diff
@@ -295,8 +332,19 @@ def compile_module(
         submodule = getattr(partitioned_module, name)
         # Criteria for a module to be convertible to TRT
         if settings.use_fast_partitioner and "_run_on_acc" not in name:
+            dryrun_tracker.to_run_in_torch.extend(parse_non_trt_nodes(submodule))
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_compiler.py:335]: - In `global_partitioning` - `named_children` only returns TRT modules'
- Can add feature for `fast_partitioning` first, then expand to global later?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 2]
```diff
@@ -341,4 +429,6 @@ def compile_module(
     if fast_partitioner_failed:
         settings.use_fast_partitioner = True
 
+    dryrun_stats_display(dryrun_tracker, settings.dryrun)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_compiler.py:432]: Pass in graph including the Torch module/node information for proper display formatting

**Pull request merged**